api/src/websocket/handlers/collab/collab.ts

```diff
@@ -0,0 +1,184 @@
+import { COLLAB, WebSocketCollabMessage, type WebSocketClient } from '@directus/types';
+import { capitalize } from 'lodash-es';
+import emitter from '../../../emitter.js';
+import { handleWebSocketError } from '../../errors.js';
+import { getMessageType } from '../../utils/message.js';
+import { CollabRooms } from './room.js';
+import type { FocusMessage, JoinMessage, LeaveMessage, SaveMessage, UpdateMessage } from './types.js';
+import { fetchAllowedCollections } from '../../../permissions/modules/fetch-allowed-collections/fetch-allowed-collections.js';
+import getDatabase from '../../../database/index.js';
+import { getSchema } from '../../../utils/get-schema.js';
+import { InvalidPayloadError } from '@directus/errors';
+import { hasFieldPermision } from './field-permissions.js';
+
+/**
+ * Handler responsible for subscriptions
+ */
+export class CollabHandler {
+	// storage of subscriptions per collection
+	rooms: CollabRooms;
+	/**
+	 * Initialize the handler
+	 */
+	constructor() {
+		this.rooms = new CollabRooms();
+		this.bindWebSocket();
+	}
+
+	/**
+	 * Hook into websocket client lifecycle events
+	 */
+	bindWebSocket() {
+		// listen to incoming messages on the connected websockets
+		emitter.onAction('websocket.message', async ({ client, message }) => {
+			if (getMessageType(message) !== COLLAB) return;
+
+			try {
+				const validMessage = WebSocketCollabMessage.parse(message);
+				await this[`on${capitalize(validMessage.action)}`](client, message);
+			} catch (error) {
+				handleWebSocketError(client, error, 'subscribe');
+			}
+		});
+
+		// unsubscribe when a connection drops
+		emitter.onAction('websocket.error', ({ client }) => this.onLeave(client));
+		emitter.onAction('websocket.close', ({ client }) => this.onLeave(client));
+	}
+
+	async onJoin(client: WebSocketClient, message: JoinMessage) {
+		try {
+			const schema = await getSchema();
+
+			const allowedCollections = await fetchAllowedCollections(
+				{ accountability: client.accountability!, action: 'read' },
+				{ knex: getDatabase(), schema },
+			);
```

Comment:
We should perform an item-level check when a users join the room. Thoughts of using `item.readOne` to confirm the item read access? 🤔